### ISSUE TITLE
Unable to compile with GCC 11.1.0

### ISSUE THREAD

**Issue by sevu** [freeorion/freeorion#3460]
There are several errors and warnings. This is from the first 25% of the compiling progress.
```
/tmp/makepkg/freeorion-git/src/freeorion-git/GG/src/GUI.cpp: In member function 'virtual void GG::GUI::PreRender()':
/tmp/makepkg/freeorion-git/src/freeorion-git/GG/src/GUI.cpp:1701:21: warning: loop variable 'modal_wnd' creates a copy from type 'const std::pair<std::shared_ptr<GG::Wnd>, std::weak_ptr<GG::Wnd> >' [-Wrange-loop-construct]
 1701 |     for (const auto modal_wnd : m_impl->m_modal_wnds) {
      |                     ^~~~~~~~~

/tmp/makepkg/freeorion-git/src/freeorion-git/GG/src/GUI.cpp:1701:21: note: use reference type to prevent copying
 1701 |     for (const auto modal_wnd : m_impl->m_modal_wnds) {
      |                     ^~~~~~~~~
      |                     &

/tmp/makepkg/freeorion-git/src/freeorion-git/GG/src/GUI.cpp: In member function 'virtual void GG::GUI::Render()':
/tmp/makepkg/freeorion-git/src/freeorion-git/GG/src/GUI.cpp:1733:21: warning: loop variable 'modal_wnd' creates a copy from type 'const std::pair<std::shared_ptr<GG::Wnd>, std::weak_ptr<GG::Wnd> >' [-Wrange-loop-construct]
 1733 |     for (const auto modal_wnd : m_impl->m_modal_wnds) {
      |                     ^~~~~~~~~

/tmp/makepkg/freeorion-git/src/freeorion-git/GG/src/GUI.cpp:1733:21: note: use reference type to prevent copying
 1733 |     for (const auto modal_wnd : m_impl->m_modal_wnds) {
      |                     ^~~~~~~~~
      |                     &
```
```
In file included from /usr/include/c++/11.1.0/bits/move.h:57,
                 from /usr/include/c++/11.1.0/bits/stl_pair.h:59,
                 from /usr/include/c++/11.1.0/bits/stl_algobase.h:64,
                 from /usr/include/c++/11.1.0/bits/stl_tree.h:63,
                 from /usr/include/c++/11.1.0/map:60,
                 from /tmp/makepkg/freeorion-git/src/freeorion-git/combat/CombatEvents.h:5,
                 from /tmp/makepkg/freeorion-git/src/freeorion-git/combat/CombatEvents.cpp:1:
/usr/include/c++/11.1.0/type_traits: In instantiation of 'struct std::is_destructible<GameRules>':
/usr/include/c++/11.1.0/future:769:21:   required from 'class std::future<GameRules>'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:31:20:   required from 'union boost::optional_detail::aligned_storage<std::future<GameRules> >::dummy_u'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:34:7:   required from 'class boost::optional_detail::aligned_storage<std::future<GameRules> >'
/usr/include/boost/optional/optional.hpp:775:18:   required from 'class boost::optional_detail::optional_base<std::future<GameRules> >'
/usr/include/boost/optional/optional.hpp:850:7:   required from 'class boost::optional<std::future<GameRules> >'
/tmp/makepkg/freeorion-git/src/freeorion-git/combat/../universe/../util/Pending.h:39:41:   required from 'struct Pending::Pending<GameRules>'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:31:20:   required from 'union boost::optional_detail::aligned_storage<Pending::Pending<GameRules> >::dummy_u'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:34:7:   required from 'class boost::optional_detail::aligned_storage<Pending::Pending<GameRules> >'
/usr/include/boost/optional/optional.hpp:775:18:   required from 'class boost::optional_detail::optional_base<Pending::Pending<GameRules> >'
/usr/include/boost/optional/optional.hpp:850:7:   required from 'class boost::optional<Pending::Pending<GameRules> >'
/tmp/makepkg/freeorion-git/src/freeorion-git/combat/../util/GameRules.h:153:83:   required from here
/usr/include/c++/11.1.0/type_traits:849:52: error: static assertion failed: template argument must be a complete class or an unbounded array
  849 |       static_assert(std::__is_complete_or_unbounded(__type_identity<_Tp>{}),
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~


/usr/include/c++/11.1.0/type_traits:849:52: note: 'std::__is_complete_or_unbounded<std::__type_identity<GameRules> >((std::__type_identity<GameRules>{}, std::__type_identity<GameRules>()))' evaluates to false
In file included from /tmp/makepkg/freeorion-git/src/freeorion-git/combat/../universe/../util/Pending.h:10,
                 from /tmp/makepkg/freeorion-git/src/freeorion-git/combat/../universe/Universe.h:21,
                 from /tmp/makepkg/freeorion-git/src/freeorion-git/combat/CombatEvents.cpp:3:
/usr/include/c++/11.1.0/future: In instantiation of 'class std::future<GameRules>':
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:31:20:   required from 'union boost::optional_detail::aligned_storage<std::future<GameRules> >::dummy_u'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:34:7:   required from 'class boost::optional_detail::aligned_storage<std::future<GameRules> >'
/usr/include/boost/optional/optional.hpp:775:18:   required from 'class boost::optional_detail::optional_base<std::future<GameRules> >'
/usr/include/boost/optional/optional.hpp:850:7:   required from 'class boost::optional<std::future<GameRules> >'
/tmp/makepkg/freeorion-git/src/freeorion-git/combat/../universe/../util/Pending.h:39:41:   required from 'struct Pending::Pending<GameRules>'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:31:20:   required from 'union boost::optional_detail::aligned_storage<Pending::Pending<GameRules> >::dummy_u'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:34:7:   required from 'class boost::optional_detail::aligned_storage<Pending::Pending<GameRules> >'
/usr/include/boost/optional/optional.hpp:775:18:   required from 'class boost::optional_detail::optional_base<Pending::Pending<GameRules> >'
/usr/include/boost/optional/optional.hpp:850:7:   required from 'class boost::optional<Pending::Pending<GameRules> >'
/tmp/makepkg/freeorion-git/src/freeorion-git/combat/../util/GameRules.h:153:83:   required from here
/usr/include/c++/11.1.0/future:769:21: error: static assertion failed: result type must be destructible
  769 |       static_assert(is_destructible<_Res>{},
      |                     ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/11.1.0/future:769:21: note: 'std::is_destructible<GameRules>{}' evaluates to false


In file included from /usr/include/c++/11.1.0/bits/move.h:57,
                 from /usr/include/c++/11.1.0/bits/stl_pair.h:59,
                 from /usr/include/c++/11.1.0/bits/stl_algobase.h:64,
                 from /usr/include/c++/11.1.0/bits/char_traits.h:39,
                 from /usr/include/c++/11.1.0/string:40,
                 from /tmp/makepkg/freeorion-git/src/freeorion-git/parse/../util/Logger.h:5,
                 from /tmp/makepkg/freeorion-git/src/freeorion-git/parse/../util/OptionsDB.h:5,
                 from /tmp/makepkg/freeorion-git/src/freeorion-git/parse/../util/GameRules.h:4,
                 from /tmp/makepkg/freeorion-git/src/freeorion-git/parse/GameRulesParser.cpp:1:
/usr/include/c++/11.1.0/type_traits: In instantiation of 'struct std::is_destructible<GameRules>':
/usr/include/c++/11.1.0/future:769:21:   required from 'class std::future<GameRules>'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:31:20:   required from 'union boost::optional_detail::aligned_storage<std::future<GameRules> >::dummy_u'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:34:7:   required from 'class boost::optional_detail::aligned_storage<std::future<GameRules> >'
/usr/include/boost/optional/optional.hpp:775:18:   required from 'class boost::optional_detail::optional_base<std::future<GameRules> >'
/usr/include/boost/optional/optional.hpp:850:7:   required from 'class boost::optional<std::future<GameRules> >'
/tmp/makepkg/freeorion-git/src/freeorion-git/parse/../util/Pending.h:39:41:   required from 'struct Pending::Pending<GameRules>'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:31:20:   required from 'union boost::optional_detail::aligned_storage<Pending::Pending<GameRules> >::dummy_u'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:34:7:   required from 'class boost::optional_detail::aligned_storage<Pending::Pending<GameRules> >'
/usr/include/boost/optional/optional.hpp:775:18:   required from 'class boost::optional_detail::optional_base<Pending::Pending<GameRules> >'
/usr/include/boost/optional/optional.hpp:850:7:   required from 'class boost::optional<Pending::Pending<GameRules> >'
/tmp/makepkg/freeorion-git/src/freeorion-git/parse/../util/GameRules.h:153:83:   required from here
/usr/include/c++/11.1.0/type_traits:849:52: error: static assertion failed: template argument must be a complete class or an unbounded array
  849 |       static_assert(std::__is_complete_or_unbounded(__type_identity<_Tp>{}),
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~


/usr/include/c++/11.1.0/type_traits:849:52: note: 'std::__is_complete_or_unbounded<std::__type_identity<GameRules> >((std::__type_identity<GameRules>{}, std::__type_identity<GameRules>()))' evaluates to false
In file included from /tmp/makepkg/freeorion-git/src/freeorion-git/parse/../util/Pending.h:10,
                 from /tmp/makepkg/freeorion-git/src/freeorion-git/parse/../util/GameRules.h:5,
                 from /tmp/makepkg/freeorion-git/src/freeorion-git/parse/GameRulesParser.cpp:1:
/usr/include/c++/11.1.0/future: In instantiation of 'class std::future<GameRules>':
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:31:20:   required from 'union boost::optional_detail::aligned_storage<std::future<GameRules> >::dummy_u'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:34:7:   required from 'class boost::optional_detail::aligned_storage<std::future<GameRules> >'
/usr/include/boost/optional/optional.hpp:775:18:   required from 'class boost::optional_detail::optional_base<std::future<GameRules> >'
/usr/include/boost/optional/optional.hpp:850:7:   required from 'class boost::optional<std::future<GameRules> >'
/tmp/makepkg/freeorion-git/src/freeorion-git/parse/../util/Pending.h:39:41:   required from 'struct Pending::Pending<GameRules>'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:31:20:   required from 'union boost::optional_detail::aligned_storage<Pending::Pending<GameRules> >::dummy_u'
/usr/include/boost/optional/detail/optional_aligned_storage.hpp:34:7:   required from 'class boost::optional_detail::aligned_storage<Pending::Pending<GameRules> >'
/usr/include/boost/optional/optional.hpp:775:18:   required from 'class boost::optional_detail::optional_base<Pending::Pending<GameRules> >'
/usr/include/boost/optional/optional.hpp:850:7:   required from 'class boost::optional<Pending::Pending<GameRules> >'
/tmp/makepkg/freeorion-git/src/freeorion-git/parse/../util/GameRules.h:153:83:   required from here
/usr/include/c++/11.1.0/future:769:21: error: static assertion failed: result type must be destructible
  769 |       static_assert(is_destructible<_Res>{},
      |                     ^~~~~~~~~~~~~~~~~~~~~~~


/usr/include/c++/11.1.0/future:769:21: note: 'std::is_destructible<GameRules>{}' evaluates to false
/usr/include/c++/11.1.0/future: In instantiation of 'class std::shared_future<GameRules>':
/usr/include/c++/11.1.0/future:788:66:   required from 'std::future<_Res>::future(std::future<_Res>&&) [with _Res = GameRules]'
/usr/include/boost/optional/optional.hpp:409:8:   required from 'void boost::optional_detail::optional_base<T>::construct(boost::optional_detail::optional_base<T>::rval_reference_type) [with T = std::future<GameRules>; boost::optional_detail::optional_base<T>::rval_reference_type = std::future<GameRules>&&]'
/usr/include/boost/optional/optional.hpp:211:18:   required from 'boost::optional_detail::optional_base<T>::optional_base(boost::optional_detail::optional_base<T>&&) [with T = std::future<GameRules>]'
/usr/include/boost/optional/optional.hpp:975:5:   required from 'Pending::Pending<T>::Pending(Pending::Pending<T>&&) [with T = GameRules]'
/usr/include/boost/optional/optional.hpp:409:8:   required from 'void boost::optional_detail::optional_base<T>::construct(boost::optional_detail::optional_base<T>::rval_reference_type) [with T = Pending::Pending<GameRules>; boost::optional_detail::optional_base<T>::rval_reference_type = Pending::Pending<GameRules>&&]'
/usr/include/boost/optional/optional.hpp:211:18:   required from 'boost::optional_detail::optional_base<T>::optional_base(boost::optional_detail::optional_base<T>&&) [with T = Pending::Pending<GameRules>]'
/usr/include/boost/optional/optional.hpp:975:5:   required from here
/usr/include/c++/11.1.0/future:906:21: error: static assertion failed: result type must be destructible
  906 |       static_assert(is_destructible<_Res>{},
      |                     ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/11.1.0/future:906:21: note: 'std::is_destructible<GameRules>{}' evaluates to false
```

**Comment by geoffthemedio** [comment]
@o01eg I suppose it could help to set up a CI build that uses GCC 11.1 ? Can we do that in Ubuntu 18 or 20?

**Comment by o01eg** [comment]
g++-11 available only on Ubuntu 21.04 or 21.10. I suppose we can use them in docker.

**Comment by o01eg** [comment]
Same error about `GameRules` on `MSVC 19.29.30037.0`

```
2021-05-31T09:04:26.6456188Z C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30037\include\type_traits(736,28): error C2139: 'GameRules': an undefined class is not allowed as an argument to compiler intrinsic type trait '__is_destructible' (compiling source file D:\a\freeorion\freeorion\parse\GameRulesParser.cpp) [D:\a\freeorion\freeorion\build\freeorionparseobj.vcxproj]
2021-05-31T09:04:26.6462006Z D:\a\freeorion\freeorion\parse\../util/GameRules.h(25): message : see declaration of 'GameRules' (compiling source file D:\a\freeorion\freeorion\parse\GameRulesParser.cpp) [D:\a\freeorion\freeorion\build\freeorionparseobj.vcxproj]
2021-05-31T09:04:26.6562951Z C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30037\include\future(880): message : see reference to variable template 'const bool is_destructible_v<GameRules>' being compiled (compiling source file D:\a\freeorion\freeorion\parse\GameRulesParser.cpp) [D:\a\freeorion\freeorion\build\freeorionparseobj.vcxproj]
2021-05-31T09:04:26.6685576Z C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30037\include\future(880,56): error C2338: T in future<T> must meet the Cpp17Destructible requirements (N4878 [futures.unique.future]/4). (compiling source file D:\a\freeorion\freeorion\parse\GameRulesParser.cpp) [D:\a\freeorion\freeorion\build\freeorionparseobj.vcxproj]
```

**Comment by o01eg** [comment]
@geoffthemedio I've setup CI to test it https://github.com/freeorion/freeorion/pull/3472

**Comment by geoffthemedio** [comment]
> Same error about `GameRules` on `MSVC`

I'm working in refactoring so the parser returns a different container, more like other content parsers...

**Comment by geoffthemedio** [comment]
> Same error about `GameRules` on `MSVC 19.29.30037.0`

Can you just use 19.28 until it's fixed, then?

**Comment by o01eg** [comment]
I suppose only the newest MSVS installed on CI so I couldn't choose older one.

**Comment by geoffthemedio** [comment]
Based on the CI build, seems to be fixed as of at least 1d8bb0dc6e7d5db95ea8ecb85f23568819ddf2b6
